include/engine/api/match_parameters_tidy.hpp

```diff
@@ -157,7 +183,22 @@ inline Result tidy(const MatchParameters &params, Thresholds cfg = {15., 5})
             if (!params.timestamps.empty())
                 result.parameters.timestamps.push_back(params.timestamps[i]);
         }
+        else
+        {
+            // one of the coordinates meant to be used as a waypoint was marked for removal
+            // update the original waypoint index to the new representative coordinate
+            const auto last_idx = result.parameters.coordinates.size() - 1;
+            if (result.was_waypoint[i] && (result.parameters.waypoints.back() != last_idx))
+            {
+                result.parameters.waypoints.push_back(last_idx);
+                // result.was_waypoint[i] = false;
+            }
+        }
     }
+    // todo: this is not always true... e.g. if the first or last coordinate is duplicated and then
+    // tidied away
+    // but we do want some kind of check like this
+    // BOOST_ASSERT(result.parameters.waypoints.size() == params.waypoints.size());
```

Comment:
Hrm I don't understand this comment or the assertion. Can we specify this more precisely?

Reply:
Hm, I actually don't think this needs to be assertion anymore, since it's not always true. I think having test coverage over this is good enough for now.